requirements-spark3.txt

```diff
@@ -25,15 +25,15 @@ backcall==0.2.0
     # via ipython
 bcrypt==3.2.0
     # via paramiko
-black==19.10b0
+black==19.10b0; extra == "all-spark3"
```

Comment:
Oh so you're running into this too.  do you know how to fix this?  I've spent at least a few hours already looking into this.  For me, it usually adds `extra == "all"` in the regular file.  Somehow, our internal qualifier is getting appended onto black's.  It makes no sense to me.

Reply:
maybe @jeevb & @cosmicBboy too if you've seen this... basically running `make requirements` adds this "extra" thing to black (only ever to black, nothing else), which then causes pip to fail because this has nothing to do with black.  "all" and "all-spark3" are flytekit concepts.

Reply:
I see. I didn't notice this. Will look into it. Thanks for pointing out.

Reply:
It seems to happen "randomly" in any compiled file, and sometimes it doesn't happen at all. :thinking:

Reply:
yeah i've had the same experience.

Reply:
After debugging deeply into pip-compile and pip, I think pip-compile got confused by `blacked` introduced by both `papermill` and `flytekit` itself as part of notebook extra. pip-compile does a few rounds of resolution until nothing changed after a round, but the order of things appearing may result differently, so sometimes it choose the one without extra marker but sometimes it doesn't.

It's weird that papermill requires black as installation requirement. Maybe we can ask the maintainer about it.

On the hand, when the extra marker gets added to black, how does pip fail in your case? To me, when there is this marker, it is as if black were not added to requirements.txt when doing `pip install -r` because I'm not providing an extra, so it doesn't fail but may happily install a newer version of black.

Maybe to avoid dealing with black and having to pin the version. Let's just unpin it and do a big reformatting PR? It doesn't change too much as I tried.

Reply:
Hmm, it actually has real code dependency on black https://github.com/nteract/papermill/blob/ca46ca6756ff15c1d03857af911f8060e35da8ab/papermill/translators.py#L189

Reply:
Down to do a reformatting yeah.  Thanks for investigating!  Do you mind putting it in?  My failure scenario... I don't remember actually.  I pulled this branch and ran `make requirements` and it ran fine.  If I see it again before the reformat I'll add it here.